code/game/objects/effects/effect_system/effects_foam.dm

```diff
@@ -139,23 +142,31 @@
 	for(var/obj/O in range(0,src))
 		if(O.type == src.type)
 			continue
+		if(O in applied_atoms && one_apply_per_object)
+			continue
 		if(isturf(O.loc))
 			var/turf/T = O.loc
 			if(T.intact && O.level == 1) //hidden under the floor
 				continue
 		if(lifetime % reagent_divisor)
 			reagents.reaction(O, VAPOR, fraction)
+			applied_atoms += O
+		CHECK_TICK
 	var/hit = 0
 	for(var/mob/living/L in range(0,src))
 		hit += foam_mob(L)
+		CHECK_TICK
 	if(hit)
 		lifetime++ //this is so the decrease from mobs hit and the natural decrease don't cumulate.
 	var/T = get_turf(src)
-	if(lifetime % reagent_divisor)
-		reagents.reaction(T, VAPOR, fraction)
+	if(!(T in applied_atoms && one_apply_per_object))
+		if(lifetime % reagent_divisor)
+			reagents.reaction(T, VAPOR, fraction)
+			applied_atoms += T
```

Comment:
check if one_apply_per_object before doing this, no need to drag a list if we have one_apply_per_object = FALSE